presto-native-execution/presto_cpp/main/operators/UnsafeRowExchangeSource.cpp

```diff
@@ -32,44 +33,63 @@ folly::SemiFuture<UnsafeRowExchangeSource::Response>
 UnsafeRowExchangeSource::request(
     uint32_t /*maxBytes*/,
     uint32_t /*maxWaitSeconds*/) {
-  std::vector<velox::ContinuePromise> promises;
-  int64_t totalBytes = 0;
   {
-    std::lock_guard<std::mutex> l(queue_->mutex());
-    if (atEnd_) {
-      return folly::makeFuture(Response{0, true});
-    }
-
-    bool hasNext;
-    CALL_SHUFFLE(hasNext = shuffle_->hasNext(), "hasNext");
-
-    if (!hasNext) {
-      atEnd_ = true;
-      queue_->enqueueLocked(nullptr, promises);
-    } else {
-      velox::BufferPtr buffer;
-      CALL_SHUFFLE(buffer = shuffle_->next(), "next");
-      totalBytes = buffer->size();
+    auto hasNext = [this]() -> folly::SemiFuture<Response> {
+      return shuffle_->hasNext()
+          .deferValue(
+              [this](bool hasNextValue) mutable
+              -> UnsafeRowExchangeSource::Response {
+                if (hasNextValue) {
+                  velox::BufferPtr buffer = shuffle_->next();
+                  int64_t totalBytes = 0;
+                  if (buffer != nullptr && buffer->size() > 0) {
+                    std::vector<velox::ContinuePromise> promises;
+                    {
+                      std::lock_guard<std::mutex> l(queue_->mutex());
+                      totalBytes = buffer->size();
+                      ++numBatches_;
 
-      ++numBatches_;
+                      auto ioBuf = folly::IOBuf::wrapBuffer(
+                          buffer->as<char>(), buffer->size());
+                      // NOTE: SerializedPage's onDestructionCb_ captures one
+                      // reference on 'buffer' to keep its alive until
+                      // SerializedPage destruction. Also note that 'buffer'
+                      // should have been allocated from memory pool. Hence, we
+                      // don't need to update the memory usage counting for the
+                      // associated 'ioBuf' attached to SerializedPage on
+                      // destruction.
+                      queue_->enqueueLocked(
+                          std::make_unique<velox::exec::SerializedPage>(
+                              std::move(ioBuf), [buffer](auto& /*unused*/) {}),
+                          promises);
+                    }
+                    for (auto& promise : promises) {
```

Comment:
Please make sure promises set is not under the lock. Thanks!

Reply:
Lock is scoped to not include setting the promise, thanks!